kvdb/postgres/readwrite_bucket.go

```diff
@@ -0,0 +1,441 @@
+				"VALUES($1, $2) RETURNING id",
+			b.id, key,
+		).Scan(&id)
+		if err != nil {
+			return nil, err
+		}
+
+	case err == nil && value != nil:
+		return nil, walletdb.ErrIncompatibleValue
+
+	case err != nil:
+		return nil, err
+	}
+
+	return newReadWriteBucket(b.tx, b.table, &id), nil
+}
+
+// DeleteNestedBucket deletes the nested bucket and its sub-buckets
+// pointed to by the passed key. All values in the bucket and sub-buckets
+// will be deleted as well.
+func (b *readWriteBucket) DeleteNestedBucket(key []byte) error {
+	if len(key) == 0 {
+		return walletdb.ErrBucketNotFound
+	}
+
+	result, err := b.tx.Exec(
+		"DELETE FROM "+b.table+" WHERE "+parentSelector(b.id)+
+			" AND key=$1 AND value IS NULL",
+		key,
+	)
+	if err != nil {
+		return err
+	}
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rows == 0 {
+		return walletdb.ErrBucketNotFound
+	}
+
+	return nil
+}
+
+// Put updates the value for the passed key.
+// Returns ErrKeyRequred if te passed key is empty.
+func (b *readWriteBucket) Put(key, value []byte) error {
+	if len(key) == 0 {
+		return walletdb.ErrKeyRequired
+	}
+
+	// Prevent NULL being written for an empty value slice.
+	if value == nil {
+		value = []byte{}
+	}
+
+	var (
+		result sql.Result
+		err    error
+	)
+	// We are putting a value in a bucket in this table. Try to insert the
+	// key first. If the key already exists (ON CONFLICT), update the key.
+	// Do not update a NULL value, because this indicates that the key
+	// contains a sub-bucket. This case will be caught via RowsAffected
+	// below.
+
+	if b.id == nil {
+		// ON CONFLICT requires the WHERE parent_id IS NULL hint to let
+		// Postgres find the NULL-parent_id unique index (<table>_unp).
+		result, err = b.tx.Exec(
+			"INSERT INTO "+b.table+" (key, value) VALUES($1, $2) "+
+				"ON CONFLICT (key) WHERE parent_id IS NULL "+
+				"DO UPDATE SET value=$2 "+
+				"WHERE "+b.table+".value IS NOT NULL",
+			key, value,
+		)
+	} else {
+		// ON CONFLICT requires the WHERE parent_id NOT IS NULL hint to
+		// let Postgres find the non-NULL-parent_id unique index
+		// (<table>_up).
+		result, err = b.tx.Exec(
+			"INSERT INTO "+b.table+" (key, value, parent_id) "+
+				"VALUES($1, $2, $3) "+
+				"ON CONFLICT (key, parent_id) "+
+				"WHERE parent_id IS NOT NULL "+
+				"DO UPDATE SET value=$2 "+
+				"WHERE "+b.table+".value IS NOT NULL",
+			key, value, b.id,
+		)
+	}
+	if err != nil {
+		return err
+	}
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if rows != 1 {
+		return walletdb.ErrIncompatibleValue
+	}
+
+	return nil
+}
+
+// Delete deletes the key/value pointed to by the passed key.
+// Returns ErrKeyRequred if the passed key is empty.
+func (b *readWriteBucket) Delete(key []byte) error {
+	if key == nil {
+		return nil
+	}
+	if len(key) == 0 {
+		return walletdb.ErrKeyRequired
+	}
+
+	// Check to see if a bucket with this key exists.
+	var dummy int
+	result := b.tx.QueryRow(
```

Comment:
`result` isn't used any further. Any reason for not using the `QueryRow().Scan()` pattern directly?

Reply:
Better indeed, fixed.